File: smash-ranks-android/app/src/test/java/com/garpr/android/misc/IdentityManagerTest.java
package com.garpr.android.misc;

import com.garpr.android.BaseTest;
import com.garpr.android.BuildConfig;
import com.garpr.android.models.AbsPlayer;
import com.google.gson.Gson;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import javax.inject.Inject;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class IdentityManagerTest extends BaseTest {

    private static final String JSON_LITE_PLAYER = "{\"id\":\"583a4a15d2994e0577b05c74\",\"name\":\"homemadewaffles\"}";

    private AbsPlayer mPlayer;

    @Inject
    Gson mGson;

    @Inject
    IdentityManager mIdentityManager;

    @Inject
    RegionManager mRegionManager;


    @Before
    @Override
    public void setUp() throws Exception {
        super.setUp();
        getTestAppComponent().inject(this);

        mPlayer = mGson.fromJson(JSON_LITE_PLAYER, AbsPlayer.class);
    }

    @Test
    public void testAddListener() throws Exception {
        final AbsPlayer[] array = new AbsPlayer[1];

        final IdentityManager.OnIdentityChangeListener listener =
                new IdentityManager.OnIdentityChangeListener() {
            @Override
            public void onIdentityChange(final IdentityManager identityManager) {
                array[0] = identityManager.getIdentity();
            }
        };

        mIdentityManager.addListener(listener);
        assertNull(array[0]);

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertEquals(mPlayer, array[0]);

        mIdentityManager.removeIdentity();
        assertNull(array[0]);
    }

    @Test
    public void testGetAndSetIdentity() throws Exception {
        assertNull(mIdentityManager.getIdentity());

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertNotNull(mIdentityManager.getIdentity());
        assertEquals(mIdentityManager.getIdentity(), mPlayer);

        mIdentityManager.removeIdentity();
        assertNull(mIdentityManager.getIdentity());
    }

    @Test
    public void testHasIdentity() throws Exception {
        assertFalse(mIdentityManager.hasIdentity());

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertTrue(mIdentityManager.hasIdentity());

        mIdentityManager.removeIdentity();
        assertFalse(mIdentityManager.hasIdentity());
    }

    @Test
    public void testIsIdWithNull() throws Exception {
        assertFalse(mIdentityManager.isId(null));

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertFalse(mIdentityManager.isId(null));

        mIdentityManager.removeIdentity();
        assertFalse(mIdentityManager.isId(null));
    }

    @Test
    public void testIsIdWithPlayer() throws Exception {
        assertFalse(mIdentityManager.isId(mPlayer.getId()));

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertTrue(mIdentityManager.isId(mPlayer.getId()));

        mIdentityManager.removeIdentity();
        assertFalse(mIdentityManager.isId(mPlayer.getId()));
    }

    @Test
    public void testIsPlayerWithNull() throws Exception {
        assertFalse(mIdentityManager.isPlayer(null));

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertFalse(mIdentityManager.isPlayer(null));

        mIdentityManager.removeIdentity();
        assertFalse(mIdentityManager.isPlayer(null));
    }

    @Test
    public void testIsPlayerWithPlayer() throws Exception {
        assertFalse(mIdentityManager.isPlayer(mPlayer));

        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertTrue(mIdentityManager.isPlayer(mPlayer));

        mIdentityManager.removeIdentity();
        assertFalse(mIdentityManager.isPlayer(mPlayer));
    }

    @Test
    public void testRemoveListener() throws Exception {
        final AbsPlayer[] array = new AbsPlayer[1];

        final IdentityManager.OnIdentityChangeListener listener =
                new IdentityManager.OnIdentityChangeListener() {
            @Override
            public void onIdentityChange(final IdentityManager identityManager) {
                array[0] = identityManager.getIdentity();
            }
        };

        mIdentityManager.addListener(listener);
        mIdentityManager.setIdentity(mPlayer, mRegionManager.getRegion());
        assertEquals(mPlayer, array[0]);

        mIdentityManager.removeListener(listener);
        mIdentityManager.removeIdentity();
        assertEquals(mPlayer, array[0]);
    }

}

File: smash-ranks-android/app/src/test/java/com/garpr/android/misc/IdentityManagerTest.kt
package com.garpr.android.misc

import com.garpr.android.BaseTest
import com.garpr.android.models.AbsPlayer
import com.google.gson.Gson
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import javax.inject.Inject

@RunWith(RobolectricTestRunner::class)
class IdentityManagerTest : BaseTest() {

    private lateinit var litePlayer: AbsPlayer
    private lateinit var rankedPlayer: AbsPlayer

    @Inject
    protected lateinit var gson: Gson

    @Inject
    protected lateinit var identityManager: IdentityManager

    @Inject
    protected lateinit var regionManager: RegionManager


    companion object {
        private const val JSON_LITE_PLAYER = "{\"id\":\"583a4a15d2994e0577b05c74\",\"name\":\"homemadewaffles\"}"
        private const val JSON_RANKED_PLAYER = "{\"rating\":41.565775187219025,\"name\":\"NMW\",\"rank\":3,\"previous_rank\":3,\"id\":\"583a4a15d2994e0577b05c8a\"}"
    }

    @Before
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        testAppComponent.inject(this)

        litePlayer = gson.fromJson(JSON_LITE_PLAYER, AbsPlayer::class.java)
        rankedPlayer = gson.fromJson(JSON_RANKED_PLAYER, AbsPlayer::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun testAddListener() {
        var identity: AbsPlayer? = null

        val listener = object : IdentityManager.OnIdentityChangeListener {
            override fun onIdentityChange(identityManager: IdentityManager) {
                identity = identityManager.identity
            }
        }

        identityManager.addListener(listener)
        assertNull(identity)

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertEquals(litePlayer, identity)

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertEquals(litePlayer, identity)

        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertEquals(rankedPlayer, identity)

        identityManager.removeIdentity()
        assertNull(identity)
    }

    @Test
    @Throws(Exception::class)
    fun testGetIdentity() {
        assertNull(identityManager.identity)
    }

    @Test
    @Throws(Exception::class)
    fun testGetAndSetIdentity() {
        assertNull(identityManager.identity)

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertEquals(litePlayer, identityManager.identity)

        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertEquals(rankedPlayer, identityManager.identity)

        identityManager.removeIdentity()
        assertNull(identityManager.identity)
    }

    @Test
    @Throws(Exception::class)
    fun testHasIdentity() {
        assertFalse(identityManager.hasIdentity)

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertTrue(identityManager.hasIdentity)

        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertTrue(identityManager.hasIdentity)

        identityManager.removeIdentity()
        assertFalse(identityManager.hasIdentity)
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithEmptyString() {
        assertFalse(identityManager.isPlayer(""))

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertFalse(identityManager.isPlayer(""))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer(""))
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithIdString() {
        assertFalse(identityManager.isPlayer(litePlayer.id))
        assertFalse(identityManager.isPlayer(rankedPlayer.id))

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertTrue(identityManager.isPlayer(litePlayer.id))
        assertFalse(identityManager.isPlayer(rankedPlayer.id))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer(litePlayer.id))
        assertFalse(identityManager.isPlayer(rankedPlayer.id))
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithNullPlayer() {
        assertFalse(identityManager.isPlayer(null as AbsPlayer?))

        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertFalse(identityManager.isPlayer(null as AbsPlayer?))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer(null as AbsPlayer?))
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithNullString() {
        assertFalse(identityManager.isPlayer(null as String?))

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertFalse(identityManager.isPlayer(null as String?))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer(null as String?))
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithPlayer() {
        assertFalse(identityManager.isPlayer(litePlayer))

        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertFalse(identityManager.isPlayer(litePlayer))
        assertTrue(identityManager.isPlayer(rankedPlayer))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer(litePlayer))
        assertFalse(identityManager.isPlayer(rankedPlayer))
    }

    @Test
    @Throws(Exception::class)
    fun testIsPlayerWithWhitespaceString() {
        assertFalse(identityManager.isPlayer(" "))

        identityManager.setIdentity(litePlayer, regionManager.getRegion())
        assertFalse(identityManager.isPlayer("  "))

        identityManager.removeIdentity()
        assertFalse(identityManager.isPlayer("   "))
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveListener() {
        var identity: AbsPlayer? = null

        val listener = object : IdentityManager.OnIdentityChangeListener {
            override fun onIdentityChange(identityManager: IdentityManager) {
                identity = identityManager.identity
            }
        }

        identityManager.addListener(listener)
        identityManager.setIdentity(rankedPlayer, regionManager.getRegion())
        assertEquals(rankedPlayer, identity)

        identityManager.removeListener(listener)
        identityManager.removeIdentity()
        assertEquals(rankedPlayer, identity)
    }

}
